### ISSUE TITLE
Don't add event as global when browser: true

### ISSUE THREAD

**Issue by kileras** [eslint/eslint#3964]
Right now when you configure your env with "browser: true" you will get a event variable automatically added to your globals so this would lint perfectly despite event not being defined (tested in the online demo):

```
/*eslint-env browser */

function a() {
  var b = event.target;
  return b;
}

a();
```

I think that this is incorrect as the event variable is only defined inside an event context, so its not totally global and may end up preventing eslint form catching some potential undefined variables.

Also, and the actual reason why I discovered this, in some contexts the event parameter and the event global may not be the same thing. For example in React they are using some fake nodes to dispatch events (I find it a bit odd but this or another similar code may still be a valid use case) and they do a bind to override the event value as seen here: https://github.com/facebook/react/blob/8c75e792abf54f34b7a36a777826f104d1e0fb34/src/shared/utils/ReactErrorUtils.js#L68 . In that case if you use the global event instead of the parameter you will get a fake <react></react> node instead of the target and you won't figure out until one hour later :smile:

For that cases if you don't think that removing the global event for the browsers is a good idea, potentially for compatibility reasons, maybe adding an specific rule to forbid the use of global event variables would be a good idea so eslint can handle this situations a bit better.

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by michaelficarra** [comment]
> I think that this is incorrect as the event variable is only defined inside an event context

Not true, it is always there. Run `"event" in this` in your console.

**Comment by kileras** [comment]
Ok fair enough, I didn't realise that it exists but is undefined :smile: 

So instead I created an issue for a new rule to disable the use of the global event variable.

**Comment by kileras** [comment]
Hi again,

I actually have the opportunity to check this again as I had the feeling that I was forgetting something about the event global and I found out that is actually not a global, or at least it is not in a consistent way. So basically event is a global variable in Webkit/Blink and Trident engines, but it is not in Gecko so in firefox if you do:
```
event in this
```
It will failed and said that event is not defined.

You will have event as a global in firefox only is you have an inline event listener, which actually makes sense and its a better behaviour as it was the reason why it was created in first place:
```
<body onclick="console.log(event);">
```
Will actually show an event as you can see in this jsbin http://jsbin.com/xuhoqibanu/edit?html,console,output

So I will say that adding event as a global is the right thing to do as its not true in all the browsers :smile:

**Comment by michaelficarra** [comment]
Wow, fringe legacy web features are weird...

**Comment by ilyavolodin** [comment]
@kileras Just to confirm, are you saying `event` should be added to globals or shouldn't be?

**Comment by Lexicality** [comment]
Bear in mind that in certain versions of IE (I think <9) `window.event` is the *only* way to access the current event.  
This doesn't matter if you're using jQuery, but if you're mucking about with events manually it's more important.

**Comment by kileras** [comment]
@ilyavolodin I'm saying that it shouldn't be added since thats not true for all the browsers. If someone needs it for some specific reason then they can add it themselves but it shouldn't be a default.

@Lexicality I'm afraid that's not actually true, I think you may be confusing the fact that IE prior to version 9 didn't have addEventListener and you need to use attachEvent, but that method also receives an event as the first parameter so there is no need to use the global event.

**Comment by nzakas** [comment]
@kileras not quite true. If you attach an event handler using DOM 0 style (onclick=) in IE <9, you still need window.event. Since this is present in nearly every browser, I'm not sure I see a good argument for removing it.

**Comment by ilyavolodin** [comment]
I think I agree with @nzakas 3 out of 4 major browsers have it as default. Unless you are coding specifically for firefox, linter should be able to notify you of issues that are related to global `event`.

**Comment by mysticatea** [comment]
I think there is a problem, we cannot remove (overwrite) env's global variables in .eslintrc. On the other hand, we can add globals.

BTW, IE<9 will expire after 3 months.

**Comment by nzakas** [comment]
That's true, however, this global exists in most browsers, so it's safer to assume that it's there by default.

**Comment by kileras** [comment]
@nzakas I know that inline event listeners depend on this, in fact I explained before that is actually the origin of it, and thats true not only for IE < 9 but for every single browser out there. That does not mean that "the only way to access the current event." is using the global as IE have its own attachEvent and you can even do, again as any other browsers since almost forever, element.onclick and then you can take it for the argument so you have choice and any of them should be taken before even thinking about using DOM 0 style events.

@ilyavolodin What you mean coding specifically for Firefox? Is is the exact opposite, this rule is specifically for Chrome/Safari and IE, which are the only browsers that support the feature. You cannot make linting rules that apply to a subset of the browsers, thats awful and completely destroy the point of eslint, which is linting javascript in general, not the specific implementation of the spec that Chrome or IE have.

On top of that I seriously hope that nobody is using DOM 0 nowadays, and it shouldn't be the default behaviour support that eslint gets when you define "browser: true" as that use of browsers events should be totally discouraged so it's a good thing to have a default that prevent you to do that.

But even more important is the fact is that, as @mysticatea said, even if you want to use them you have an option which is to add event to your globals, which is pretty easy and painful, but you cannot do the opposite, you cannot say I don't want to use this global variable. That alone makes the case for not adding it to the variables as I explained before this can shadow important errors and behaviours that can take a long time to figure out.

**Comment by kileras** [comment]
I just saw that is not exactly this repo the one that has the list of globals and in fact someone already created and issue and a pr to fix it so I drop a comment there too for reference :smile:

**Comment by zowers** [comment]
having `event` in globals and no means to exclude it is the source of "event is undefined" errors, so there are 2 possibilities:
1. exclude `event` from `globals.browser`
2. have a way to exclude variables defined earlier by `env` configuration

**Comment by nzakas** [comment]
@kileras we are not bound to use whatever `globals` exposes, we still need to define for ourselves whether or not we want to include it. So far, I'm not seeing a benefit of removing `event`. Unless you want to show how having `event` predefined is causing harm, I'm inclined to say we aren't making this change. (Leaving myself open to be convinced here, but the argument that Firefox doesn't support it isn't enough.)

**Comment by kileras** [comment]
@nzakas Well for me there is not much difference between window.event, which is a proprietary implementation present in two engines and not part of any spec, and the prefixed globals with webkit for example, they are equally specific and I think that neither of them should be considered global. 

But if you don't feel like it's enough, well its your decision :smile:  Feel free to close then issue I'll continue the conversation on https://github.com/eslint/eslint/issues/3966 and see if we can get some kind of rule to prevents its use.

**Comment by Lexicality** [comment]
@kileras Apparently if you use `element.onclick = myfunc`, [myfunc will not be passed the event in old IE](http://www.quirksmode.org/js/events_access.html#link3), necessitating the use of `window.event`.

**Comment by kileras** [comment]
@Lexicality By the look of it that article seems to be quite old, it doesn't even has any reference to attachEvent/addEventListener and talks about Netscape, so I would say that the IE version that they are talking about its very old. I don't remember since it has been a while, luckily :smile: , but I think that at least IE6 passes the event in that case so we would be talking here about IE 5.5 or less which is old enough to not be taken into consideration and remove it even from our thoughts :smile:  (and if you ask me IE6 too at this point).

**Comment by nzakas** [comment]
Thanks for the good discussion. We aren't going to remove `event` at this time.

**Comment by rjbailey** [comment]
I was nearly burned by this. Came close to shipping a change where some handlers for jQuery delegated events were mistakenly written taking no parameters, instead of one `event` parameter. It worked fine everywhere other than Firefox, of course.

Isn't it safer to reject code that *might* be safe (because it relies on a non-standard global variable) rather than always accepting code that is potentially unsafe in a widely used browser?

Projects that rely on the `event` global can always declare it in their `globals` configuration, so removing `event` from the globals list won't prevent projects from using it, if they need to. The way things are now, I don't see a way to prevent this failure from happening again other than disabling the `browser` environment and declaring the globals I need manually, which is less than ideal.

Would it be possible to offer a way of taking the `browser` globals but omitting `event`?

**Comment by zowers** [comment]
@rjbailey I made a fork, installed globals from it, removed globals from eslint/node_modules
https://github.com/metacommunications/globals/tree/fix-remove-event

**Comment by mzgol** [comment]
You might all be interested in my `no-global-event` proposal: #4710.

**Comment by rhys-vdw** [comment]
Just throwing in my 2c here. Global event fails in both modern versions of IE and Firefox (but passes in Chrome). Since I do all my development in chrome this one just slipped by ESLint and (thankfully) ended up being caught in a PR review:

```js
  onMouseDown() { // <-- missing `event` argument
    if (event.button === 0 && event.target === this.$element.get(0)) {
      this.startTranslate(event);
    }
  }
```

I'd certainly prefer a linter error if an undefined `event` exists in scope. My take is that if you're doing this deliberately, then you can add manually the `event` global to your config (or just reference it explicitly as `window.event`). Alternatively offer some additional configuration option like `browser-legacy` or similar?

Seems silly to allow code that will break in modern browsers in preference for permitting code that is *only* required by old browsers.

**Comment by kileras** [comment]
@rhys-vdw Well as the one I opened the ticket originally I totally agree 😄  .... but its kind of fixed now with the implementation of "no-restricted-globals" which basically allow you to prevent the use of any globals. So if you do:
```
no-restricted-globals: [2, "event"]
```
Then in theory (to be honest I didn't test it yet) any uses of event as global will show an error in eslint. So basically a new default setting in all my eslint config files 😛

**Comment by rhys-vdw** [comment]
@kileras Yeah, I discovered that afterwards. I realized afterwards there are a host of things attached to `window` that I don't want in every scope. `name` being an example.

So my solution

```js
// ...
  "env": {
    "browser": false
  },
  "globals": {
    "process": true,
    "window": true,
    "console": true,
    "document": true,
    "setInterval": true,
    "setTimeout": true
  }
```

If you want to use `alert` it needs to be `window.alert`.

In short I don't think that `env.browser` is a useful option.

**Comment by axelson** [comment]
Note, you can use the `no-restricted-globals` rule now: http://eslint.org/docs/rules/no-restricted-globals
